File: 134-gas-station/134-gas-station.py
class Solution:
    def canCompleteCircuit(self, gas: List[int], cost: List[int]) -> int:
        start, n = 0, len(gas) - 1
        visited = dict()
        while(len(visited) <= n):
            cur = 0
            i = start
            while(cur + gas[i] >= cost[i]):
                visited[i] = 1
                cur += gas[i] - cost[i]
                i += 1
                i = i%(n+1)
                if i == start:
                    return start
            visited[i] = 1
            start = i + 1
        return -1
            
                
                

File: 134-gas-station/134-gas-station.java
class Solution {
    public int canCompleteCircuit(int[] gas, int[] cost) {
        //@input : gas/cost from i to i + 1
        //@output : the starting position such that it can go through all the elements
        //@edge case : integer limit, negative number
        //@breaking down problem:
        //1. calculate the difference
        //2. calculate the prefix sum for the differences
        //3. traverse the difference and calcu
        //注意此时有解且唯一，说明满足条件即可,相减和大于等于0，说明必有解
        int n = gas.length;
        int[] diff = new int[n];
        int sum = 0;
        int curStreak = 0;
        int ans = 0;
        for(int i = 0; i < n; i++){
            diff[i] = gas[i] - cost[i];
            curStreak += diff[i];
            if(curStreak < 0){
                curStreak = 0;
                ans = i + 1;
            }
            sum += diff[i];
        }
        if(sum < 0)return -1;
        return ans;
        
        
        
        
        
    }
}